### PR TITLE
DEVPROD-4768: bulk write updates when caching host data from EC2

### PR DESCRIPTION
**Pull request by Kimchelly**
DEVPROD-4768

### Description
The host status job is responsible for polling host statuses so that Evergreen knows when they're up and running. When [getting the host status](https://github.com/evergreen-ci/evergreen/blob/99807e9b0813b44eb7c07958b8f1cbc4858f77e3/units/host_status.go#L104), it also [caches some of the information it gets](https://github.com/evergreen-ci/evergreen/blob/99807e9b0813b44eb7c07958b8f1cbc4858f77e3/cloud/ec2_util.go#L305-L327). This caching is done via many `UpdateOne` statements, so if the job checks `N` hosts and finds they're running, it will do approximately `2N` `UpdateOne`s sequentially.

Based on [Honeycomb stats](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen/result/z5owHjDpkbH), the host status job occasionally takes several minutes to complete, and it looks like it's taking a long time to serially run the many `UpdateOne`s to cache the data. In [this example](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen/result/z5owHjDpkbH/trace/RknW3hhoAn?fields[]=s_name&fields[]=s_serviceName&span=d331f3c4458fbff7), it took 1 second to get the statuses from EC2 but 5 minutes to finish all the `UpdateOne`s. During periods when Evergreen is starting up lots of hosts, this bottleneck means hosts may take a long time to get checked.

This PR reduces the amount of roundtrips to the DB by [bulk writing the updates to the hosts](https://www.mongodb.com/docs/manual/core/bulk-write-operations/#bulk-write-operations), which should speed things up when there's a lot of updates. So instead of doing `2N` individual trips to the DB to do the updates, it now sends all the cache updates in bulk to the DB in one trip.

I'm going to monitor the performance change after this is deployed. Depending on how much this speeds up the job, there may be a follow-up PR to further optimize it.

* Bulk write updates to cache EC2 instance data in host documents.
* Consolidate IPv6 and public IPv4 address updates so they get updated with the other EC2 instance data when caching.
* Refactor caching logic to consolidate common logic between caching 1 host vs. caching many.
* Simplify down some overly complicated logic in EC2 implementation for getting instance statuses.

### Testing
* Updated existing tests and added a couple more for better coverage.
* Tested in staging by spawning a spawn host and starting a fresh EC2 task host. Verified that both started up with all their EC2 instance data cached.

### Documentation
N/A

### CODE REVIEW

File: cloud/ec2.go [hunk 1]
```diff
@@ -709,46 +709,60 @@ func (m *ec2Manager) GetInstanceStatuses(ctx context.Context, hosts []host.Host)
...
+				host:     instanceIdToHostMap[hostID],
+				instance: &instance,
+			})
+			hostIDsToCache = append(hostIDsToCache, hostID)
 		}
+		hostToStatusMap[hostID] = status
 	}
 
+	// Cache instance information so we can make fewer calls to AWS's API.
+	grip.Error(message.WrapError(cacheAllHostData(ctx, m.env, m.client, hostsToCache), message.Fields{
```
Comment (hadjri) [inline cloud/ec2.go:760]: Why not return an error on failure as we were doing before?

Comment (hadjri) [inline cloud/ec2.go:760]: Why not return an error on failure here as we were doing before?

File: cloud/ec2_util.go [hunk 2]
```diff
@@ -329,6 +314,85 @@ func cacheHostData(ctx context.Context, env evergreen.Environment, h *host.Host,
...
+	}
+
+	catcher.Wrap(host.CacheAllCloudProviderData(ctx, env, hostsToCache), "bulk caching host data")
+
+	return catcher.Resolve()
+}
+
+func makeCloudProviderData(instance *types.Instance) (*host.CloudProviderData, error) {
+	if err := validateCloudProviderData(instance); err != nil {
+		return nil, err
```
Comment (hadjri) [inline cloud/ec2_util.go:362]: Can we wrap this with a message saying validation failed?

File: cloud/ec2_util.go [hunk 3]
```diff
@@ -282,40 +282,25 @@ func validateUserDataSize(userData, distroID string) error {
...
-	}
-	if instance.LaunchTime == nil {
-		return errors.New("instance missing launch time")
-	}
-	if instance.PublicDnsName == nil {
-		return errors.New("instance missing public DNS name")
-	}
-	if instance.PrivateIpAddress == nil {
-		return errors.New("instance missing private IP address")
+func cacheHostData(ctx context.Context, env evergreen.Environment, client AWSClient, pair hostInstancePair) error {
```
Comment (hadjri) [inline cloud/ec2_util.go:292]: I understand the DB optimizations between single vs multi insert are why this and `cacheAllHostData` are distinct, but it seems some of the logic between these two can be shared

File: cloud/ec2_util.go [hunk 4]
```diff
@@ -282,40 +282,25 @@ func validateUserDataSize(userData, distroID string) error {
...
+	h := pair.host
+	instance := pair.instance
+	data, err := makeCloudProviderData(instance)
+	if err != nil {
+		return errors.Wrap(err, "making cloud provider data")
```
Comment (hadjri) [inline cloud/ec2_util.go:297]: nit: `catcher.Wrapf(err, "making cloud provider data for host '%s'", h.Id)`

File: model/host/host.go [hunk 5]
```diff
@@ -1893,25 +1872,74 @@ func (h *Host) Upsert(ctx context.Context) (*mongo.UpdateResult, error) {
...
+	Zone        string
+	StartedAt   time.Time
+	PublicDNS   string
+	PublicIPv4  string
+	PrivateIPv4 string
+	IPv6        string
+	Volumes     []VolumeAttachment
+}
+
+// CacheCloudProvider caches data about the host from its cloud provider.
```
Comment (hadjri) [inline model/host/host.go:1887]: nit: CacheCloudProviderData

**Review by hadjri — CHANGES_REQUESTED** [review]

File: cloud/ec2_util.go (hunk 2 above)
Comment (Kimchelly) [reply cloud/ec2_util.go]: Done.

File: model/host/host.go (hunk 5 above)
Comment (Kimchelly) [reply model/host/host.go]: This has been removed.

**Review by hadjri — APPROVED** [review]

**Pull request merged**